File: dp/coin change.js
/**
 * Leetcode 322. Coin Change
 * https://leetcode.com/problems/coin-change/
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    
    let dp = Array(amount+5).fill(Infinity)
    dp[0] = 0;
    
    for(let coin of coins){
        for(let x=coin ; x<=amount ; x++){
            dp[x] = Math.min(dp[x], 1+ dp[x - coin])
        }
    }

    return dp[amount] == Infinity? -1: dp[amount];

};

File: dp/coin change.py
class Solution:

    def run(self, amount, coins, memo):
        if amount == 0:
            return 0
        if amount < 0:
            return 10**9

        if amount in memo:
            return memo[amount]

        best = 10**9
        for c in coins:
            best = min(best, 1 + self.run(amount - c, coins, memo))

        memo[amount] = best
        return best

    def coinChange(self, coins: List[int], amount: int) -> int:
        memo = {}
        ans = self.run(amount, coins, memo)
        return ans if ans != 10**9 else -1